src/main/java/com/google/api/codegen/py/PythonGapicContext.java

```diff
@@ -296,11 +321,88 @@ private String returnTypeComment(
     return pythonCommon.convertToCommentedBlock(contentBuilder.toString());
   }
 
+  private String methodSample(Method method, PythonImportHandler importHandler) {
+    if (!getApiConfig().generateSamples()) {
+      return "";
+    }
+
+    Interface service = (Interface) method.getParent();
+    String apiName = getApiWrapperName(service);
+    List<String> importStrings = new ArrayList<>();
+    importStrings.add(apiImport(apiName));
+    String methodName = upperCamelToLowerUnderscore(method.getSimpleName());
+    String returnType = returnTypeOrEmpty(method.getOutputType(), importHandler);
+    MethodConfig methodConfig = getApiConfig().getInterfaceConfig(service).getMethodConfig(method);
+    List<Field> fields = Lists.newArrayList(methodConfig.getRequiredFields());
+
+    PythonDocConfig docConfig =
+        PythonDocConfig.newBuilder()
+            .setApiName(apiName)
+            .setAppImports(importStrings)
+            .setMethodName(methodName)
+            .setReturnType(returnType)
+            .setFieldInitCode(this, service, method, fields)
+            .setFieldParams(this, fields)
+            .setIterableResponse(methodConfig.isPageStreaming())
+            .build();
+
+    protoImports(method, docConfig);
+    return generateMethodSampleCode(docConfig);
+  }
+
+  private String returnTypeOrEmpty(TypeRef returnType, PythonImportHandler importHandler) {
+    return messages().isEmptyType(returnType)
+        ? ""
+        : typeCardinalityComment(returnType, importHandler);
+  }
+
+  private String apiImport(String apiName) {
+    String packageName = getApiConfig().getPackageName();
+    String moduleName = packageName + "." + lowerCamelToLowerUnderscore(apiName);
+    return PythonImport.create(ImportType.APP, moduleName, apiName).importString();
+  }
+
+  // mutates appImports list in docConfig
+  private void protoImports(Method method, PythonDocConfig docConfig) {
+    List<String> importStrings = docConfig.getAppImports();
+    for (InitCodeLine line : docConfig.getInitCode().getLines()) {
+      TypeRef lineType = null;
+      switch (line.getLineType()) {
+        case SimpleInitLine:
+          lineType = ((SimpleInitCodeLine) line).getType();
+          break;
+        case StructureInitLine:
+          lineType = ((StructureInitCodeLine) line).getType();
+          break;
+        default:
+          // nothing to do
+      }
+
+      if (lineType != null && lineType.isMessage()) {
+        importStrings.addAll(new PythonImportHandler(method).calculateImports());
+        return;
+      }
+    }
+  }
+
+  /**
+   * Generate the sample code for a method.
+   */
+  public String generateMethodSampleCode(PythonDocConfig config) {
+    return pythonSnippetSet.generateMethodSampleCode(config).prettyPrint();
```

Comment:
I am a little concerned about the complexity of running snippet codegen within snippet codegen -- I think it makes the code hard to follow, and removes the ability for the sample templates to "talk" to the main.snip templates about things like indentation and formatting.

I would suggest breaking the Python comment methods in PythonGapicContext into pre- and post- sample sections so that generateMethodSampleCode can be called directly from main.snip, like in Java.

(In the long term, I think we should try to adopt a more Java-like approach to comments by leveraging the snippet code rather than doing all of the comment generation in the context, but that's for another day.)